pkg/frontend/querymiddleware/experimental_operators.go

```diff
@@ -0,0 +1,151 @@
+		return &experimentalOperatorsMiddleware{
+			next:   next,
+			limits: limits,
+			logger: logger,
+		}
+	})
+}
+
+func (m *experimentalOperatorsMiddleware) Do(ctx context.Context, req MetricsQueryRequest) (Response, error) {
+	tenantIDs, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return nil, apierror.New(apierror.TypeBadData, err.Error())
+	}
+
+	enabledExperimentalFunctions := make(map[string][]string, len(tenantIDs))
+	allExperimentalFunctionsEnabled := true
+	for _, tenantID := range tenantIDs {
+		// note that this includes both functions and aggregations (ie limitk, limit_ratio)
+		enabled := m.limits.EnabledPromQLExperimentalFunctions(tenantID)
+		enabledExperimentalFunctions[tenantID] = enabled
+		if len(enabled) == 0 || enabled[0] != allExperimentalOperators {
+			allExperimentalFunctionsEnabled = false
+		}
+	}
+
+	enabledExtendedRangeSelectors := make(map[string][]string, len(tenantIDs))
+	allExtendedRangeSelectorsEnabled := true
+	for _, tenantID := range tenantIDs {
+		enabled := m.limits.EnabledPromQLExtendedRangeSelectors(tenantID)
+		enabledExtendedRangeSelectors[tenantID] = enabled
+		if len(enabled) == 0 || enabled[0] != allExperimentalOperators {
+			allExtendedRangeSelectorsEnabled = false
+		}
+	}
+
+	if allExperimentalFunctionsEnabled && allExtendedRangeSelectorsEnabled {
+		// If all experimental functions are enabled for all tenants here, we don't need to check the query
+		// for those functions and can skip this middleware.
+		return m.next.Do(ctx, req)
+	}
+
+	expr, err := astmapper.CloneExpr(req.GetParsedQuery())
+	if err != nil {
+		return nil, apierror.New(apierror.TypeBadData, DecorateWithParamName(err, "query").Error())
+	}
+	operations := containedExperimentalOperations(expr)
+	if len(operations) == 0 {
+		// This query does not contain any experimental functions, so we can continue to the next middleware.
+		return m.next.Do(ctx, req)
+	}
+
+	// Make sure that every used experimental operation is enabled for all the tenants here.
+	var tenantMap map[string][]string
+	for op, operationType := range operations {
+		switch operationType {
+		case functionType, aggregationType:
+			tenantMap = enabledExperimentalFunctions
+		case extendedRangeSelectorModifierType:
+			tenantMap = enabledExtendedRangeSelectors
+		}
+
+		for _, enabled := range tenantMap {
+			if len(enabled) > 0 && enabled[0] == allExperimentalOperators {
+				// If the first item matches the const value of allExperimentalOperations, then all experimental
+				// operations are enabled for this tenant.
+				continue
+			}
+			if !slices.Contains(enabled, op) {
+				return nil, createExperimentalOperationError(operationType, op)
+			}
+		}
+	}
+
+	// Every used experimental operation is enabled for the tenant(s).
+	return m.next.Do(ctx, req)
+}
+
+func createExperimentalOperationError(operationType experimentalOperationType, operation string) error {
+	err := fmt.Errorf("%s \"%s\" is not enabled for tenant", operationType.label, operation)
```

Comment:
`%q` will quote the string for you, and I think we should include the word "experimental" to make the reason why clearer:

```suggestion
	err := fmt.Errorf("experimental %s %q is not enabled for tenant", operationType.label, operation)
```